crates/trie/src/state.rs

```diff
@@ -293,13 +328,32 @@ impl HashedStorage {
     /// Create new instance of [HashedStorage].
     pub fn new(wiped: bool) -> Self {
         Self {
-            non_zero_valued_storage: Vec::new(),
+            non_zero_valued_slots: Vec::new(),
             zero_valued_slots: AHashSet::new(),
             wiped,
             sorted: true, // empty is sorted
         }
     }
 
+    /// Extend hashed storage with contents of other.
+    /// The entries in second hashed storage take precedence.
+    pub fn extend(&mut self, other: Self) {
+        let mut entries: HashMap<B256, U256> =
+            HashMap::from_iter(self.non_zero_valued_slots.drain(..).chain(
+                self.zero_valued_slots.drain().map(|hashed_slot| (hashed_slot, U256::ZERO)),
+            ));
+        for (hashed_slot, value) in other.non_zero_valued_slots {
+            entries.insert(hashed_slot, value);
+        }
+        for hashed_slot in other.zero_valued_slots {
+            entries.insert(hashed_slot, U256::ZERO);
+        }
+        for (hashed_slot, value) in entries {
+            self.insert_slot(hashed_slot, value);
+        }
```

Comment:
same q w.r.t. draining

Reply:
same answer as above, not happy about, but it deserves a separate PR